### PR TITLE
Reverse Op

### PR DESCRIPTION
**Pull request by cagery**
Adds reverse op

### CODE REVIEW

File: src/targets/gpu/device/reverse.cpp [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+argument reverse(hipStream_t stream, argument result, argument arg1, int64_t axis)
+{
+    auto s                = arg1.get_shape();
+    auto axis_dim_size    = s.lens()[axis];
+    std::size_t nelements = s.elements();
+    visit_all(result, arg1)([&](auto output1, auto input1) {
+        hip_visit_views(output1, input1, s)([&](auto output, auto input, auto hs) {
+            // const auto* in_ptr = device_cast(input.data());
+            // auto* out_ptr = device_cast(output.data());
+            gs_launch(stream, nelements, 256)([=](auto i) __device__ {
```
Comment (pfultz2) [inline src/targets/gpu/device/reverse.cpp:22]: Is there a reason you need to set the local size to 256?

File: src/targets/gpu/device/reverse.cpp [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+argument reverse(hipStream_t stream, argument result, argument arg1, int64_t axis)
+{
+    auto s                = arg1.get_shape();
+    auto axis_dim_size    = s.lens()[axis];
+    std::size_t nelements = s.elements();
+    visit_all(result, arg1)([&](auto output1, auto input1) {
+        hip_visit_views(output1, input1, s)([&](auto output, auto input, auto hs) {
+            // const auto* in_ptr = device_cast(input.data());
+            // auto* out_ptr = device_cast(output.data());
```
[the remark this answers is not in the document]
Comment (cagery) [reply src/targets/gpu/device/reverse.cpp]: Yes, removed.

File: test/ref_ops_test.cpp [hunk 3]
```diff
@@ -3560,6 +3560,46 @@ TEST_CASE(reshape_test)
...
+    migraphx::program p;
+    auto* mm = p.get_main_module();
+    auto l   = mm->add_literal(migraphx::literal{in_shape, data});
+    int axis = 1;
+    mm->add_instruction(migraphx::make_op("reverse", {{"axis", axis}}), l);
+    p.compile(migraphx::ref::target{});
+    auto result = p.eval({}).back();
+    std::vector<float> results_vector;
+    result.visit([&](auto output) { results_vector.assign(output.begin(), output.end()); });
+    std::vector<float> target_data = {16, 15, 14, 13, 12, 11, 10, 9,  8,  7,  6,
```
Comment (kahmed10) [inline test/ref_ops_test.cpp:3577]: It might be easier to make a copy of the original data, then call std::reverse on it. This way if the test changes in the future, it will be easier to manage.

Comment (cagery) [reply test/ref_ops_test.cpp:3577]: What kind of test changes do you refer to?

Comment (kahmed10) [reply test/ref_ops_test.cpp:3577]: right now, you have the target_data values hardcoded. I'm suggesting to generate these numbers and call std::reverse on them.

Comment (cagery) [reply test/ref_ops_test.cpp:3577]: What is the benefit? Hardcoded is less computation vs generating copy and reversing it is additional computation. I am not seeing clearly the benefit of it.

Comment (kahmed10) [reply test/ref_ops_test.cpp:3577]: It's preference, but if the test changes or becomes more complex, you would have to redo the values. The other option is to just reverse an existing vector in the expected behavior. I'm okay either way for this PR.

Comment (pfultz2) [reply test/ref_ops_test.cpp:3577]: Using `std::reverse` would help ensure the data is correct, whereas hardcoded data may not always be obvious.

File: src/targets/gpu/device/reverse.cpp (hunk 1 above)
Comment (pfultz2) [reply src/targets/gpu/device/reverse.cpp:20]: Yea, defaulting to 1024 should better utilize the GPU. If you remove the parameter it will default to 1024.

File: test/ref_ops_test.cpp (hunk 3 above)
Comment (cagery) [reply test/ref_ops_test.cpp:3577]: I understand but then with the same idea, we should not have had any hardcoded gold vectors in these tests, e.g. round_test or pow_test, there are many of them, for which we have standard functions to use.

File: src/targets/gpu/device/reverse.cpp (hunk 1 above)
Comment (cagery) [reply src/targets/gpu/device/reverse.cpp:20]: Okay makes sense, will replace it accordingly. Related, why does gather op use 256, any specific reason ?https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/blob/0a9c18a582721085b211abb842c5b77c0fcce12b/src/targets/gpu/device/gather.cpp#L27) ?

File: test/ref_ops_test.cpp (hunk 3 above)
Comment (cagery) [reply test/ref_ops_test.cpp]: Replaced gold vectors with `std::reverse` and `std::swap_ranges` depending on axis. Retested, looks ok.

File: src/targets/gpu/device/reverse.cpp (hunk 1 above)
Comment (cagery) [reply src/targets/gpu/device/reverse.cpp]: Changed it to 1024

File: test/ref_ops_test.cpp (hunk 3 above)
Comment (kahmed10) [reply test/ref_ops_test.cpp]: > I understand but then with the same idea, we should not have had any hardcoded gold vectors in these tests, e.g. round_test or pow_test, there are many of them, for which we have standard functions to use.

Right, I think moving forward once we refactor other tests, we should replace the hardcoded values there as well.

**Review by kahmed10 — APPROVED** [review]

File: src/targets/gpu/device/reverse.cpp (hunk 1 above)
Comment (scxiao) [reply src/targets/gpu/device/reverse.cpp]: @cagery For whatever reason, 1024 for gather does not work correctly.

File: test/verify/test_reverse.cpp [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    {
+        migraphx::program p;
+        auto* mm = p.get_main_module();
+        migraphx::shape s{migraphx::shape::float_type, {4, 16}};
+        auto a0  = mm->add_parameter("data", s);
+        int axis = 0;
+        mm->add_instruction(migraphx::make_op("reverse", {{"axis", axis}}), a0);
+        return p;
+    }
+};
```
Comment (scxiao) [inline test/verify/test_reverse.cpp:18]: Maybe you can add a test for axis to be a negative number.

**Review by pfultz2 — APPROVED** [review]

**Review by scxiao — APPROVED** [review]

File: test/verify/test_reverse.cpp (hunk 4 above)
Comment (cagery) [reply test/verify/test_reverse.cpp:18]: Added `test/verify/test_reverse_negaxis.cpp`
